File: pca_cluster.R
###################################
# PCA and cluster analysis
###############################
# setup
#install.packages("factoextra")
#install.packages("cluster")
library(tidyverse)
library(factoextra)
library(cluster)

setwd("C:/Users/stepa/downloads")
df<-read_csv("0_kw_analysis.csv")

###
# define the variables of interest
filenames <- df$filename
features <- df %>% select(-filename)

# run PCA
pca_result <- prcomp(features, scale. = TRUE)
summary(pca_result)
fviz_eig(pca_result, 
  title = "Scree Plot of Principal Components",
  xlab = "Principal Components",
  ylab = "Percentage of explained variances")

# feature importance in components
loadings <- data.frame(pca_result$rotation)
top_loadings <- loadings %>%
  mutate(feature = rownames(loadings)) %>%
  gather(PC, loading, -feature) %>%
  group_by(PC) %>%
  arrange(desc(abs(loading))) %>%
  slice_head(n = 5)

top_loadings

# PC1 and 2
pc12_loadings<-top_loadings %>%
  filter(PC %in% c("PC1", "PC2")) %>%
  arrange(PC, desc(abs(loading)))
pc12_loadings



# PCA biplot
fviz_pca_biplot(pca_result,
  geom.ind = "point",  
  label = "var",      
  title = "PCA of Paper Concepts")

###################################
# cluster analysis
set.seed(456)
n_clusters <- 5  
km_result <- kmeans(pca_result$x[, 1:2], centers = n_clusters)

# visualize the clusters
fviz_cluster(km_result, 
             data = pca_result$x[, 1:2],
             main = "PCA with K-means Clustering")

######################
# Random Forest
install.packages("randomForest")
library(randomForest)

target <- features$`political science`
predictors <- features %>% 
  select(-`political science`)

# enter the (random) forest!
set.seed(123)
rf_model <- randomForest(x = predictors, 
                         y = target,
                         importance = TRUE)

# variable importance
importance_df <- as.data.frame(importance(rf_model)) %>%
  rownames_to_column("feature") %>%
  arrange(desc(`%IncMSE`))

# plot top 20 most important features
ggplot(head(importance_df, 20), 
       aes(x = reorder(feature, `%IncMSE`), y = `%IncMSE`)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  theme_minimal() +
  labs(x = "Features",
       y = "% Increase in MSE",
       title = "Top 20 Concepts Most Predictive of Political Science")

File: pca_cluster.py
###################################
# PCA and cluster analysis
###############################
# setup
pip install scikit-learn
pip install matplotlib
pip install seaborn
import pandas as pd
import numpy as np
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
import matplotlib.pyplot as plt
import seaborn as sns

# Load data
dp = pd.read_csv("C:/Users/stepa/Downloads/0_kw_analysis.csv")

###
# define the variables of interest
filenames = dp['filename']
features = dp.drop('filename', axis=1)

# Run PCA
pca = PCA()
features_scaled = StandardScaler().fit_transform(features)
pca_result = pca.fit_transform(features_scaled)
print("Explained variance ratio:", pca.explained_variance_ratio_[:5])

# Scree plot
plt.figure(figsize=(10, 6))
plt.plot(range(1, len(pca.explained_variance_ratio_) + 1), 
         pca.explained_variance_ratio_, 'bo-')
plt.title('Scree Plot of Principal Components')
plt.xlabel('Principal Components')
plt.ylabel('Percentage of explained variances')
plt.show()

# Feature importance in components
loadings = pd.DataFrame(
    pca.components_.T,
    columns=[f'PC{i+1}' for i in range(len(pca.components_))],
    index=features.columns
)

# top 5 features per component
top_loadings = pd.DataFrame()
for pc in loadings.columns:
    top_5 = pd.DataFrame({
        'feature': loadings.index,
        'PC': pc,
        'loading': loadings[pc]
    })
    # Sort by absolute values manually
    top_5 = top_5.reindex(top_5['loading'].abs().sort_values(ascending=False).index)
    top_5 = top_5.head(5)
    top_loadings = pd.concat([top_loadings, top_5])



print(top_loadings)


# PC1 and 2
pc12_loadings = top_loadings[top_loadings['PC'].isin(['PC1', 'PC2'])]
pc12_loadings = pc12_loadings.sort_values(['PC', 'loading'], 
                                         ascending=[True, False])
print(pc12_loadings)

# PCA biplot
plt.figure(figsize=(12, 8))
plt.scatter(pca_result[:, 0], pca_result[:, 1], alpha=0.5)

for i, feature in enumerate(features.columns):
    plt.arrow(0, 0, 
              pca.components_[0, i]*max(abs(pca_result[:, 0])),
              pca.components_[1, i]*max(abs(pca_result[:, 1])),
              color='r', alpha=0.5)
    plt.text(pca.components_[0, i]*max(abs(pca_result[:, 0]))*1.15,
             pca.components_[1, i]*max(abs(pca_result[:, 1]))*1.15,
             feature)

plt.xlabel('First Principal Component')
plt.ylabel('Second Principal Component')
plt.title('PCA of Paper Concepts')
plt.show()

###################################
# cluster analysis
np.random.seed(456)
n_clusters = 5
kmeans = KMeans(n_clusters=n_clusters)
clusters = kmeans.fit_predict(pca_result[:, :2])

# visualize clusters
plt.figure(figsize=(10, 8))
scatter = plt.scatter(pca_result[:, 0], pca_result[:, 1], 
                     c=clusters, cmap='viridis')
plt.colorbar(scatter, label='Cluster')
plt.title('PCA with K-means Clustering')
plt.xlabel('First Principal Component')
plt.ylabel('Second Principal Component')
plt.show()